Initial attempt to define Lagrange multiplier variables in Cartesian system

dewenyushu opened the pull request:
Addresses #21461

dewenyushu committed:
Initial change to the contact constraints

Need MooseVariable type to have things working in MortarConstraintBase::zeroInactiveLMDofs

Revert changes in ADMortarConstraint, make variables of the proper type

Project LMs in ComputeWeightedGapLMMechanicalContact

Add tangential constraints

dewenyushu committed:
use processResidualAndJacobian for 3 different constraints

recuero committed:
Avoid the creation of upstream Lagrange multipliers to use a Cartesian definition to perform mechanical contact. That is, general mortar constraints do not (hopefully) need to be aware of what or how we are doing things in the contact module.

Assign normal 2D contact, nodal normals/tangent vectors for solving normal complementarity conditions with Cartesian LMs.

One bouncing block test appears to be working fine (mortar_cartesian_lms folder)

(#21461)

dewenyushu commented:
ping @recuero

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1097827/) on 933bb4a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 580898752edde8036c172e4d0f4b77b4eec9242f`

recuero commented on framework/include/constraints/MortarConstraintBase.h:
```diff
@@ -102,6 +102,7 @@ class MortarConstraintBase : public Constraint,
   /// Reference to the finite element problem
   FEProblemBase & _fe_problem;
 
+protected:
```
We shouldn't need `protected` here.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1097888/) on 4587082 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21462/site)

This comment will be updated on new commits.

dewenyushu committed:
Modify examples for cartesian lm variables. The cases converge using LU. However, while using AMG, linear iteration is still highly dependent on the contact interface angle.

dewenyushu committed:
Switch between dof x or y depending on normal direction

dewenyushu committed:
Add more input cases for comparison

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1101601/) on 9d764b1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a4d6ceb558ee7c425777f208ac92fc54242cb9ec`

recuero committed:
Add 3D Cartesian LMs

- Also create a criterion for selecting LMs associated with constraints
  that depends on last time step's normal (to avoid changing the Jacobian within
  one time step).

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1109914/) on 60baac3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d655d3a3e0548088bc3138295662d810c5f0702a`

dewenyushu committed:
Add MortarPressureComponentAux for transforming cartesian LM values to local coordinates and add test cases

recuero committed:
Add Cartesian-based mortar frictional forces (#21461)

recuero committed:
Cartesian LMs are an alternative to local contact frame LMs (#21461)

Initially we thought of fully substituting the local frame formulation by
the Cartesial LMs approach. But, due to the extent of development of local LMs'
mechanical contact, it may be more straightforward to allow both approaches to
coexist while we further test the implementation for accuracy, condensation,
usage of iterative solvers.

This PR also forces the Cartesian objects to avoid interpolating geometry. In other words,
it uses a single local frame for each constraint. This prevents deviations from orthonormal
frames while interpolating geometry within an element. This may not be the source of
a lot of error in general, so we may consider also adding the interpolation option in the
future.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1121426/) on 6d18527 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e518384dbc4ab461a1fa8e2e560c00483117102c`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1121662/) on 9348eff wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 44c5531270651773c2d51897bacef04f430410e9`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1121951/) on 17226cc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 44c5531270651773c2d51897bacef04f430410e9`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1122149/) on 9144319 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 44c5531270651773c2d51897bacef04f430410e9`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1123356/) on 152da3d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1a9cd6b0e69c310b515ce7fed5ed11461004434b`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1123424/) on 165b1d4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>1a9cd6</th>
      <th colspan=4>#21462 165b1d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.79%</td>
      <td>83.78%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>78038</td>
      <td>78034</td>
      <td>-4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15100</td>
      <td>15104</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21462/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21462/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>1a9cd6</th>
      <th colspan=4>#21462 165b1d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.98%</td>
      <td>88.78%</td>
      <td>-0.20%</td>
      <td>87.20%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2915</td>
      <td>3283</td>
      <td>+368</td>
      <td>368</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>361</td>
      <td>415</td>
      <td>+54</td>
      <td>54</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21462/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21462/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21462/coverage/xfem)

</details>

## Warnings

- `contact` new line coverage rate 87.20% is less than the suggested 90.0%



This comment will be updated on new commits.

recuero committed:
Add simple 3D friction case

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1124063/) on 58748c0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21462/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d672f863d19813cf7021e6e1a044d8e445e9a340`

recuero committed:
Avoid unused variable warning, local AD

dewenyushu committed:
Add vcp examples for frictional cases.

recuero committed:
Use tensor mechanics master action for cylinder tests

dewenyushu committed:
Fix cylinder_friction_cartesian test

dewenyushu committed:
Add tests and documentation for MortarPressureComponentAux

lindsayad reviewed (changes requested):
I would like to see in the documentation what makes these classes different from the existing variationally consistent mortar constraint classes. This would be helpful in allowing me to determine (without closely comparing the existing code) whether there should be code re-used between these new and existing classes.

recuero committed:
Consolidate parallel communication code in mortar constraints.

Provide motivation for the use of Cartesian Lagrange multipliers in the
context of mortar contact mechanics. The main reason this approach is
considered is for seamlessly using the variable condensation preconditioner,
which condenses the Lagrange multiplier based on global tractions. In order to
retrieve global tractions from the existing variationally consistent approach with
LMs based on the local frame, one needs to account for mortar geometry. This renders the
use of VCP impossible without a specific adaption of the preconditioner. For this reason,
and since it's also used in the literature within this variationally consistent approach, we
propose adding this variation in the definition of Lagrange multipliers.

lindsayad commented on modules/contact/doc/content/source/constraints/ComputeWeightedGapCartesianLMMechanicalContact.md:
```diff
@@ -44,6 +42,22 @@ gap is of order .01, then `c` should be set to 1e6 in order to bring
+Cartesian Lagrange multipliers is recommended when condensing Lagrange multipliers via the variable condensation preconditioner
+(VCP) [`VariableCondensationPreconditioner`](/VariableCondensationPreconditioner.md).
+
+The user can also employed locally oriented Lagrange multipliers `ComputeWeightedGapLMMechanicalContact`](/ComputeWeightedGapLMMechanicalContact.md),
```
```suggestion
The user can also employed locally oriented Lagrange multipliers [`ComputeWeightedGapLMMechanicalContact`](/ComputeWeightedGapLMMechanicalContact.md),
```

lindsayad commented on modules/contact/include/utils/MortarContactUtils.h:
```diff
@@ -0,0 +1,116 @@
+#include "timpi/parallel_sync.h"
+#include "timpi/communicator.h"
+
+namespace Moose
+{
+namespace Contact
+{
+namespace Mortar
```
I think I would prefer that `Mortar` be outside of `Contact` since we have `Moose::Mortar` invocations in the framework

lindsayad commented on modules/contact/doc/content/source/constraints/ComputeWeightedGapCartesianLMMechanicalContact.md:
```diff
@@ -44,6 +42,22 @@ gap is of order .01, then `c` should be set to 1e6 in order to bring
+(VCP) [`VariableCondensationPreconditioner`](/VariableCondensationPreconditioner.md).
+
+The user can also employed locally oriented Lagrange multipliers `ComputeWeightedGapLMMechanicalContact`](/ComputeWeightedGapLMMechanicalContact.md),
+which minimizes the number of contact constraints.
```
Let me know if my understanding is correct or not... if you are using the Cartesian strategy, then you *must* have frictional contact constraints, correct? Or else you could potentially have a singular row if an axis is aligned with the mortar face?

On the other hand if you are not doing the Cartesian strategy, you can optionally choose to just use a single normal contact constraint. If you choose to add frictional constraints with a non-Cartesian strategy, then you should have the same number of contact constraints as the Cartesian strategy.

Is my analysis accurate or not?

recuero replied:
>  if you are using the Cartesian strategy, then you must have frictional contact constraints, correct? 

I think what you say in general is right. But when we go with the Cartesian strategy, we will _always_ have the same number of constraints, which coincides with the dimension of the problem. You can still choose to not have frictional constraints by just imposing that tangential tractions can be zero (i.e. we _have_ to add constraints but are frictionless), which is what happens when we don't add additional constraints for the Lagrange multipliers defined in the local frame with the normal constraint. So, using the `ComputeWeightedGapCartesianLMMechanicalContact` constraint allows the user to have Cartesial LMs and frictionless behavior.

recuero replied:
> On the other hand if you are not doing the Cartesian strategy, you can optionally choose to just use a single normal contact constraint. If you choose to add frictional constraints with a non-Cartesian strategy, then you should have the same number of contact constraints as the Cartesian strategy.

👍

recuero replied:
> Or else you could potentially have a singular row if an axis is aligned with the mortar face?

There is a strategy in the contact constraint to handle the degree of freedom associated with each of the constraint. To avoid, for example, associate a normal constraint along x with a z Lagrange multiplier. This has shown to work in examples in which two blocks in contact are rotated from 0 to 90 degrees (@dewenyushu checked this), but it is intuitively to think the structure of the Jacobian will be poorer with this approach compared to the local LM approach. Sorry for the choppy answer.

lindsayad replied:
So if you want to do frictional contact with the Cartesian strategy, then in the MOOSE implementation speak you have a single contact constraint, and it is the frictional object. This single frictional object acts on the all the LMs (2 in 2D, 3 in 3D). If you want frictionless, then you use the frictionless object (`*WeightedGap*`); again this single constraint object acts on all the LMs.

This is different from the local LM approach in which we have a constraint object per Lagrange Multiplier (one weighted gap constraint object for the normal LM and 1-2 frictional constraint objects for the tangential LMs).

That about right?

recuero replied:
In the local LM approach, we have one single object for the frictional forces which may add one or two constraints depending on whether we are doing 2D or 3D. 

On the other hand, if we were speaking of adding the generalized forces corresponding to the contact constraint, with the Cartesian LM approach, we only need one object (`CartesianMortarMechanicalContact`) because there is no need to project the contribution of the LM, since it's already in the global frame.

lindsayad replied:
> In the local LM approach, we have one single object for the frictional forces which may add one or two constraints depending on whether we are doing 2D or 3D.

Oh you are right.

> On the other hand, if we were speaking of adding the generalized forces corresponding to the contact constraint, with the Cartesian LM approach, we only need one object (`CartesianMortarMechanicalContact`) because there is no need to project the contribution of the LM, since it's already in the global frame.

That's nice!

lindsayad reviewed (changes requested):
These are very nice additions/improvements

recuero commented:
This may be ready for a re-review

lindsayad commented on modules/contact/doc/content/source/auxkernels/MortarPressureComponentAux.md:
```diff
@@ -0,0 +1,24 @@
+\begin{equation}
+\boldsymbol{z}_j = (z_n)_j \boldsymbol{n}_j + (z_\tau^{1})_j \boldsymbol{\tau}^{1}_j + (z_\tau^{2})_j \boldsymbol{\tau}^{2}_j ,
+\end{equation}
+where $z_n$ is the component that is normal to the contact surface,  $\boldsymbol{n}$ is the local normal vector. For frictional problems, the $z_\tau^{1}$ is the tangential component that is aligned with the contact surface, $\boldsymbol{\tau}^{1}$ is the corresponding tangential vector. In three-dimensional frictional problems, an additional component, $z_\tau^{2}$ exist along the contact surfa ...
```
> tangential component that is aligned with the contact surface

This statement makes sense for 2D in which the contact surface is 1D. I don't really think this needs to be reworded. It's just something I thought about.

recuero replied:
Made a small change to highlight that that particular statement refers to two dimensions.

lindsayad commented on modules/contact/doc/content/source/constraints/ComputeFrictionalForceCartesianLMMechanicalContact.md:
```diff
@@ -0,0 +1,45 @@
+This class represents a preliminary implementation of frictional mortar contact constraints intended to be used with Lagrange's multiplier interpolation with dual bases. The nonlinear complementarity constraints employed here are based on a primal-dual active set strategy (PDASS), see [!citep](gitterle2010finite). These constraints capture nodes in sticking and slipping states on different soluti ...
+
+\begin{equation}
+C_{tj}(\lambda_{j},\boldsymbol{u}, \boldsymbol{\dot{u}}) = \max({\mu({pressure} + c_{n}\tilde{g}_{nj}),  abs({\lambda_{j} + c_t \tilde{u}_{tj}}))  \lambda_{j} - \mu \max({0,({pressure} + c_{n}\tilde{g}_{nj})}) (\lambda_{j} + c_t \tilde{u}_{tj})
```
Do we actually want to spell out pressure? I don't really care.

Maybe use \mathrm for abs

recuero replied:
Made those two changes. Called pressure "p". The normal LM can be introduced instead, but calling it "p" may make it more general and independent of the formulation used to compute frictional forces

lindsayad commented on modules/contact/doc/content/source/constraints/ComputeFrictionalForceCartesianLMMechanicalContact.md:
```diff
@@ -0,0 +1,45 @@
+}
+\end{equation}
+
+$\lambda_{j}$ is a Lagrange's multiplier that refers to the tangential contact pressure at node $j$, $\tilde{u}_{tj}$ is the weighted tangential velocity integrated forward in time, $\tilde{g}_n)_j$ is the weighted normal gap, $c_{n}$ is a numerical parameter ($c$ in [`ComputeWeightedGapLMMechanicalContact`](/ComputeWeightedGapLMMechanicalContact.md)) and $c_{t}$ is a numerical parameter that can ...
```
I think that we should not do `` formatting for class names that we want as links because it gets set as code and doesn't appear to the viewer as linkable

recuero replied:
Hmm It appears as linkable to me, at least, locally. Made that change throughout the mortar mechanical contact object documentation pages.

lindsayad replied:
Oh it is linkable for sure. It just doesn't appear as the standard blue hyperlink when it's set as code

lindsayad commented on modules/contact/doc/content/source/constraints/ComputeFrictionalForceCartesianLMMechanicalContact.md:
```diff
@@ -0,0 +1,45 @@
+The other degree(s) of freedom are constrained by
+enforcing that tangential tractions follow Coulomb constraints within a semi-smooth Newton approach. Usage of
+Cartesian Lagrange multipliers is recommended when condensing Lagrange multipliers via the variable condensation preconditioner
+(VCP) [`VariableCondensationPreconditioner`](/VariableCondensationPreconditioner.md).
```
```suggestion
(VCP) [VariableCondensationPreconditioner](/VariableCondensationPreconditioner.md).
```
following my comment above

lindsayad commented on modules/contact/doc/content/source/constraints/ComputeWeightedGapCartesianLMMechanicalContact.md:
```diff
@@ -0,0 +1,67 @@
+contact constraint, which, in general, will be a function of all (two or three) Cartesian Lagrange multipliers.
+This methodology only constrains one degree of freedom. The other degree(s) of freedom are constrained by
+enforcing that tangential tractions are identically zero. Note that, if friction with Cartesian Lagrange multipliers
+is chosen via [`ComputeFrictionalForceCartesianLMMechanicalContact`](/ComputeFrictionalForceCartesianLMMechanicalContact.md),
```
```suggestion
is chosen via [ComputeFrictionalForceCartesianLMMechanicalContact](/ComputeFrictionalForceCartesianLMMechanicalContact.md),
```

lindsayad commented on modules/contact/doc/content/source/constraints/ComputeWeightedGapCartesianLMMechanicalContact.md:
```diff
@@ -0,0 +1,67 @@
+is chosen via [`ComputeFrictionalForceCartesianLMMechanicalContact`](/ComputeFrictionalForceCartesianLMMechanicalContact.md),
+those remaining nodal degrees of freedom are constraint using Coulomb constraints within a semi-smooth Newton approach. Usage of
+Cartesian Lagrange multipliers is recommended when condensing Lagrange multipliers via the variable condensation preconditioner
+(VCP) [`VariableCondensationPreconditioner`](/VariableCondensationPreconditioner.md).
```
```suggestion
(VCP) [VariableCondensationPreconditioner](/VariableCondensationPreconditioner.md).
```

lindsayad commented on modules/contact/doc/content/source/constraints/ComputeWeightedGapCartesianLMMechanicalContact.md:
```diff
@@ -0,0 +1,67 @@
+Cartesian Lagrange multipliers is recommended when condensing Lagrange multipliers via the variable condensation preconditioner
+(VCP) [`VariableCondensationPreconditioner`](/VariableCondensationPreconditioner.md).
+
+The user can also employed locally oriented Lagrange multipliers [`ComputeWeightedGapLMMechanicalContact`](/ComputeWeightedGapLMMechanicalContact.md),
```
```suggestion
The user can also employed locally oriented Lagrange multipliers [ComputeWeightedGapLMMechanicalContact](/ComputeWeightedGapLMMechanicalContact.md),
```

lindsayad commented on modules/contact/include/auxkernels/MortarPressureComponentAux.h:
```diff
@@ -0,0 +1,50 @@
+
+#include "AuxKernel.h"
+
+class MortarPressureComponentAux : public AuxKernel
```
Class doxygen would be nice

recuero replied:
Added.

lindsayad commented on modules/contact/include/constraints/CartesianMortarMechanicalContact.h:
```diff
@@ -0,0 +1,25 @@
+
+#include "ADMortarLagrangeConstraint.h"
+
+class CartesianMortarMechanicalContact : public ADMortarLagrangeConstraint
```
class doxygen

lindsayad commented on modules/contact/include/constraints/CartesianMortarMechanicalContact.h:
```diff
@@ -0,0 +1,25 @@
+protected:
+  ADReal computeQpResidual(Moose::MortarType type) final;
+
+  const MooseEnum _component;
```
doxy

recuero replied:
Added

lindsayad commented on modules/contact/src/auxkernels/MortarPressureComponentAux.C:
```diff
@@ -0,0 +1,136 @@
+  Point normal_vector, tangent1, tangent2;
+
+  for (const auto lowerd_node : make_range(lower_dimensional_element->n_nodes()))
+  {
+    if (_current_node->id() == lower_dimensional_element->node_id(lowerd_node))
+    {
+      normal_vector = nodal_normals[lowerd_node];
+      tangent1 = nodal_tangents[0][lowerd_node];
+      tangent2 = nodal_tangents[1][lowerd_node];
+      break;
+    }
+  }
```
```suggestion
    if (_current_node->id() == lower_dimensional_element->node_id(lowerd_node))
    {
      normal_vector = nodal_normals[lowerd_node];
      tangent1 = nodal_tangents[0][lowerd_node];
      tangent2 = nodal_tangents[1][lowerd_node];
      break;
    }
```

lindsayad commented on modules/contact/src/constraints/CartesianMortarMechanicalContact.C:
```diff
@@ -0,0 +1,66 @@
+  switch (type)
+  {
+    case Moose::MortarType::Secondary:
+      // If normals is positive, then this residual is positive, indicating that we have an outflow
+      // of momentum, which in turn indicates that the momentum will tend to decrease at this
+      // location with time, which is what we want because the force vector is in the negative
+      // direction (always opposite of the normals). Conversely, if the normals is negative, then
+      // this residual is negative, indicating that we have an inflow of momentum, which in turn
+      // indicates the momentum will tend to increase at this location with time, which is what we
+      // want because the force vector is in the positive direction (always opposite of the
+      // normals).
+      // Get the _dof_to_weighted_gap map
+
+      return _test_secondary[_i][_qp] * _lambda[_qp];
+
+    case Moose::MortarType::Primary:
+      // The normal vector is signed according to the secondary face, so we need to introduce a
+      // negative sign here
```
These comments (which were relevant in the class this was probably copied from) regarding normals can be removed.

recuero replied:
Yeah.

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceCartesianLMMechanicalContact.C:
```diff
@@ -0,0 +1,358 @@
+    nz = _dof_to_old_normal_vector[dof](2);
+  }
+  unsigned int component_normal = 0;
+  if (std::abs(ny) > 0.57735)
```
I don't understand this number. ny could be > .57735 and nx or nz could be larger than it (this can be true up to ny = .707 = sqrt(0.5)). Where does it come from?

recuero replied:
A generic unit vector in 3D can be: {1/sqrt(3), 1/sqrt(3), 1/sqrt(3)}. Any deviation from this would result in a dominant component within the vector (1/sqrt(3) = 0.57735). As it is coded, it would be suboptimal for some cases in 2D, which I wasn't too concerned about honestly. The intention of this code is to avoid flagrant ill conditioning. Made some changes, let me know if you'd like something else.

lindsayad replied:
Even in a 3D simulation you could have a normal like {1/sqrt(2), 1/sqrt(2), 0}

recuero replied:
Yeah, I'd just go with the generic definition instead axis-aligned case. From the few simple cases I have run, this particular difference will not have a significant practical impact.

lindsayad replied:
but indeed it would be problematic if you *never* detected an "important" component, as you would not if you set the threshold to 1/sqrt(2) and then in 3D had a {1/sqrt(3), 1/sqrt(3), 1/sqrt(3)} vector. So yea I like the change you made and the variable name you gave

lindsayad commented on modules/contact/test/tests/mortar_aux_kernels/pressure-aux-friction-3d.i:
```diff
@@ -0,0 +1,372 @@
+  end_time = 1
+  dt = .5
+  dtmin = .01
+  solve_type = 'NEWTON'
```
Newton? Does it do better than PJFNK on this problem? If we're using Newton, then we can remove `-mat_mffd_err` from the petsc options

lindsayad replied:
Similarly for any other inputs with Newton and `-mat_mffd_err`

recuero replied:
Many mechanical problems with mortar contact can do as well with Newton as with PJFNK. Not all of them, but many. 
I think I removed all unused `-mat_mffd_err`.

Also, the variable condensation preconditioner has shown to work best for Newton compared to PJFNK, this is to be investigated.

lindsayad commented on modules/contact/test/tests/mortar_aux_kernels/tests:
```diff
@@ -179,4 +184,48 @@
+    min_ad_size = '50'
+    ad_indexing_type = 'global'
+    mesh_mode = 'REPLICATED'
+    requirement = 'The MortarPressureComponentAux should produce correct normal contact pressure '
```
Let's keep implementation details (like class names) out of requirements. Similarly for other requirements in this spec

recuero replied:
Done

lindsayad commented on modules/contact/test/tests/mortar_cartesian_lms/frictionless-mortar-3d-friction.i:
```diff
@@ -0,0 +1,363 @@
+  scaling = 1e0
+[]
+
+[Problem]
+  # error_on_jacobian_nonzero_reallocation = true
+[]
+
```
```suggestion
```

lindsayad commented on modules/contact/test/tests/mortar_cartesian_lms/frictionless-mortar-3d.i:
```diff
@@ -0,0 +1,361 @@
+  scaling = 1e0
+[]
+
+[Problem]
+  # error_on_jacobian_nonzero_reallocation = true
+[]
+
```
```suggestion
```

lindsayad commented on modules/contact/test/tests/mortar_cartesian_lms/frictionless-weighted-gap-lm.i:
```diff
@@ -0,0 +1,168 @@
+  [lm_y]
+    block = 3
+  []
+  # [lm_z]
+  #   block = 3
+  #   initial_condition = 0.0
+  # []
```
```suggestion
```

lindsayad commented on modules/contact/test/tests/mortar_cartesian_lms/tests:
```diff
@@ -0,0 +1,84 @@
+    max_parallel = 4
+    min_ad_size = '100'
+    ad_indexing_type = 'global'
+    mesh_mode = 'REPLICATED'
```
I know that VCP requires a replicated mesh but is that being used here?

lindsayad replied:
So this test, even though it doesn't involve the VCP, fails with distributed mesh? If so, then that sounds like a bug

recuero replied:
It does involve the variable condensation preconditioner.

lindsayad commented on modules/contact/test/tests/mortar_cartesian_lms/tests:
```diff
@@ -0,0 +1,84 @@
+    max_parallel = 4
+    min_ad_size = '100'
+    ad_indexing_type = 'global'
+    mesh_mode = 'REPLICATED'
```
Or here?

recuero replied:
Yeah, it is used. Maybe I am missing your point. But I see that this is a VCP test which is run (or tried to) with a distributed mesh.

lindsayad replied:
oh, it *is* a VCP test? Basically I want to make sure we can run all non-VCP tests with distributed mesh.

lindsayad reviewed (changes requested):
This is impressive work. I'm very impressed by the coverage. I continue to wonder whether there are more opportunities for code deduplication, but I appreciate the deduplication work done here with `MortarContactUtils`.

@dschwen commented to me once that folks worked hard to remove `Compute` prefixes from a bunch of materials and now we've added it back to the contact constraints (and I'm the one who did it). Any interest in renaming in this PR? Probably better not to actually. There is already a ton here. I'll create an issue instead.

recuero committed:
Address Alex's comments

lindsayad commented on modules/contact/test/tests/mortar_aux_kernels/tests:
```diff
@@ -179,4 +184,48 @@
+    min_ad_size = '50'
+    ad_indexing_type = 'global'
+    mesh_mode = 'REPLICATED'
+    requirement = 'The system shall produce correct normal contact pressure '
```
```suggestion
    requirement = 'The system shall produce the correct normal contact pressure '
```

lindsayad commented on modules/contact/test/tests/mortar_aux_kernels/tests:
```diff
@@ -179,4 +184,48 @@
+    min_ad_size = '100'
+    ad_indexing_type = 'global'
+    mesh_mode = 'REPLICATED'
+    requirement = 'The system shall should produce correct normal contact pressure '
```
```suggestion
    requirement = 'The system shall should produce the correct normal contact pressure '
```

lindsayad commented on modules/contact/test/tests/mortar_aux_kernels/tests:
```diff
@@ -179,4 +184,48 @@
+    min_ad_size = '50'
+    ad_indexing_type = 'global'
+    mesh_mode = 'REPLICATED'
+    requirement = 'The system shall produce correct tangential contact pressure '
```
```suggestion
    requirement = 'The system shall produce the correct tangential contact pressure '
```

lindsayad commented on modules/contact/test/tests/mortar_aux_kernels/tests:
```diff
@@ -179,4 +184,48 @@
+    min_ad_size = '100'
+    ad_indexing_type = 'global'
+    mesh_mode = 'REPLICATED'
+    requirement = 'The system shall produce correct tangential contact pressure '
```
```suggestion
    requirement = 'The system shall produce the correct tangential contact pressure '
```

lindsayad commented on modules/contact/test/tests/mortar_cartesian_lms/tests:
```diff
@@ -0,0 +1,84 @@
+    min_ad_size = '50'
+    valgrind = 'heavy'
+    ad_indexing_type = 'global'
+    requirement = 'Contact module shall solve a two-dimensional frictionless problem with Cartesian Lagrange multipliers.'
```
```suggestion
    requirement = 'The contact module shall be able to solve a two-dimensional frictionless problem with Cartesian Lagrange multipliers.'
```

lindsayad commented on modules/contact/test/tests/mortar_cartesian_lms/tests:
```diff
@@ -0,0 +1,84 @@
+    min_ad_size = '100'
+    ad_indexing_type = 'global'
+    mesh_mode = 'REPLICATED'
+    requirement = 'Contact module shall solve a three-dimensional frictionless problem with Cartesian Lagrange multipliers.'
```
```suggestion
    requirement = 'The contact module shall be able to solve a three-dimensional frictionless problem with Cartesian Lagrange multipliers.'
```

lindsayad replied:
similarly for following test specs

lindsayad reviewed (changes requested):
Very close

recuero committed:
Improve test requirement wording

lindsayad reviewed (approved):
Nice!
